go-metadata/istio.go

```diff
@@ -0,0 +1,132 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istiometadata
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// ConnectionMetadata returns the metadata about a connection
+type ConnectionMetadata struct {
+	Identity string `json:"identity"`
+}
+
+// metadataClient provides a default client for all metadata requests
+var metadataClient = http.Client{
+	Timeout: time.Second,
+}
+
+// metadataContextKey is the key the metadata handler will store connection metadata in
+var metadataContextKey = &contextKey{}
+
+type contextKey struct {
+	name string
+}
+
+func (k *contextKey) String() string { return "istio metadata context value" }
+
+func metadataServerURL() *url.URL {
+	u, err := url.Parse("http://169.254.169.111")
+	if err != nil {
+		panic(err.Error())
+	}
+	return u
+}
+
+func lookup(src, dst string) (*ConnectionMetadata, error) {
+	srcip, srcport, err := net.SplitHostPort(src)
+	if err != nil {
+		return nil, fmt.Errorf("invalid src: %v", err)
+	}
+	dstip, dstport, err := net.SplitHostPort(dst)
+	if err != nil {
+		return nil, fmt.Errorf("invalid dst: %v", err)
+	}
+
+	u := metadataServerURL()
+	params := url.Values{}
+	params.Add("srcip", srcip)
+	params.Add("srcport", srcport)
+	params.Add("dstip", dstip)
+	params.Add("dstport", dstport)
+	u.RawQuery = params.Encode()
+
+	resp, err := metadataClient.Get(u.String())
+	if err != nil {
+		return nil, fmt.Errorf("metadata lookup failed: %v", err)
+	}
+
+	mr := &ConnectionMetadata{}
+	if err := json.NewDecoder(resp.Body).Decode(mr); err != nil {
+		return nil, err
+	}
+	return mr, nil
+}
+
+// Handler is an HTTP middleware that looks up the metadata for each request, storing it in context.
+// It can later be looked up with ExtractFromRequest.
+func Handler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Context().Value(metadataContextKey) != nil {
+			// We already have a value (connection re-use), skip
+			h.ServeHTTP(w, r)
+			return
+		}
+		local := r.Context().Value(http.LocalAddrContextKey)
+		cm, err := lookup(r.RemoteAddr, local.(*net.TCPAddr).String())
+		if err != nil {
+			// Nothing we can do
+			h.ServeHTTP(w, r)
+			return
+		}
+		// Serve underlying handler
+		ctx := context.WithValue(r.Context(), metadataContextKey, cm)
+		h.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// ExtractFromRequest attempts to extract ConnectionMetadata from a request.
+// This requires Handler to be used.
+// If a connection is re-used between requests, only a single call will be made.
+func ExtractFromRequest(r *http.Request) *ConnectionMetadata {
```

Comment:
That's what I don't like about this pattern - user has to call ExtractFromRequest ( so has a dependency to this library anyways ), but the Handler is called in ALL requests. If user only needs peer info for some requests - it's more complicated to setup ( wrap only some request with the handler chain ), or not possible if the lookup is based on some other logic ( like checking XFF or other standard headers). 

Instead this method could just to the lookup on demand, and no wrapper needed.